gnucash/register/register-gnome/completioncell-gnome.c

```diff
@@ -0,0 +1,1010 @@
+        gnc_item_list_select (box->item_list, NULL);
+        unblock_list_signals (cell);
+        gnc_item_edit_hide_popup (box->item_edit);
+        box->list_popped = FALSE;
+    }
+
+    gnc_basic_cell_set_value_internal (bcell, newval);
+    *cursor_position = *start_selection = newval_chars;
+    *end_selection = -1;
+}
+
+static gboolean
+gnc_completion_cell_direct_update (BasicCell* bcell,
+                                   int* cursor_position,
+                                   int* start_selection,
+                                   int* end_selection,
+                                   void* gui_data)
+{
+    CompletionCell* cell = (CompletionCell*) bcell;
+    PopBox* box = cell->cell.gui_private;
+    GdkEventKey* event = gui_data;
+
+    if (event->type != GDK_KEY_PRESS)
+        return FALSE;
+
+    switch (event->keyval)
+    {
+    case GDK_KEY_Tab:
+    case GDK_KEY_ISO_Left_Tab:
+        {
+            char* string = gnc_item_list_get_selection (box->item_list);
+
+            if (!string)
+                break;
+
+            g_signal_emit_by_name (G_OBJECT(box->item_list), "change_item",
+                                   string, (gpointer)bcell);
+
+            g_free (string);
+            break;
+        }
+    }
+
+    if (box->strict)
+        box->in_list_select = gnc_item_in_list (box->item_list, bcell->value);
+
+    if (!bcell->value)
+        item_store_clear (cell);
+
+    return FALSE;
+}
+
+void
+gnc_completion_cell_reverse_sort (CompletionCell* cell, gboolean is_reversed)
+{
+    PopBox* box;
+
+    if (cell == NULL)
+        return;
+
+    box = cell->cell.gui_private;
+
+    if (is_reversed != box->register_is_reversed)
+    {
+        gnc_completion_cell_clear_menu (cell);
+
+        box->register_is_reversed = is_reversed;
+        box->occurrence_id = 0;
+    }
+}
+
+static void
+toggle_filter_any_cb (GtkWidget *widget, gpointer user_data)
+{
+    PopBox* box = user_data;
+    box->filter_at_start = !gtk_check_menu_item_get_active (GTK_CHECK_MENU_ITEM(widget));
+}
+
+static void
+toggle_use_alpha_cb (GtkWidget *widget, gpointer user_data)
+{
+    PopBox* box = user_data;
+
+    box->filter_sort_use_alpha = gtk_check_menu_item_get_active (GTK_CHECK_MENU_ITEM(widget));
+
+    if (box->filter_sort_use_alpha)
+        box->sort_column_id = TEXT_NORMAL_COL;
+    else
+        box->sort_column_id = OCCURANCE_COL;
+}
+
+static void
+disconnect_secondary_icon (PopBox* box)
+{
+    if (box->icon_release_id != 0)
+        g_signal_handler_disconnect (G_OBJECT(box->item_edit->editor),
+                                     box->icon_release_id);
+    box->icon_release_id = 0;
+
+    gtk_entry_set_icon_from_icon_name (GTK_ENTRY(box->item_edit->editor),
+                                       GTK_ENTRY_ICON_SECONDARY,
+                                       NULL);
+}
+
+static void
+icon_release_cb (GtkEntry *entry, GtkEntryIconPosition icon_pos,
```

Comment:
Using a GtkEntryIcon to pop a menu is a neat trick :)
Depending on where we get with our sort/filter algorithm it may no longer be necessary for this use case.

Reply:
Already scrapped.